src/Core/Session.php

```diff
@@ -66,8 +70,142 @@ public static function get($name, $defaults = null)
 		return $return;
 	}
 
+	/**
+	 * Sets a single session variable.
+	 * Overrides value of existing key.
+	 *
+	 * @param string $name
+	 * @param mixed $value
+	 */
 	public static function set($name, $value)
 	{
 		$_SESSION[$name] = $value;
 	}
+
+	/**
+	 * Sets multiple session variables.
+	 * Overrides values for existing keys.
+	 *
+	 * @param array $values
+	 */
+	public static function setMultiple(array $values)
+	{
+		$_SESSION = $values + $_SESSION;
+	}
+
+	/**
+	 * Removes a session variable.
+	 * Ignores missing keys.
+	 *
+	 * @param $name
+	 */
+	public static function remove($name)
+	{
+		unset($_SESSION[$name]);
+	}
+
+	/**
+	 * @brief Sets the provided user's authenticated session
+	 *
+	 * @param App   $a
+	 * @param array $user_record
+	 * @param bool  $login_initial
+	 * @param bool  $interactive
+	 * @param bool  $login_refresh
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	public static function setAuthenticatedForUser(App $a, array $user_record, $login_initial = false, $interactive = false, $login_refresh = false)
+	{
+		self::setMultiple([
+			'uid'           => $user_record['uid'],
+			'theme'         => $user_record['theme'],
+			'mobile-theme'  => PConfig::get($user_record['uid'], 'system', 'mobile_theme'),
+			'authenticated' => 1,
+			'page_flags'    => $user_record['page-flags'],
+			'my_url'        => $a->getBaseURL() . '/profile/' . $user_record['nickname'],
+			'my_address'    => $user_record['nickname'] . '@' . substr($a->getBaseURL(), strpos($a->getBaseURL(), '://') + 3),
+			'addr'          => defaults($_SERVER, 'REMOTE_ADDR', '0.0.0.0'),
+		]);
+
+		$member_since = strtotime($user_record['register_date']);
+		self::set('new_member', time() < ($member_since + ( 60 * 60 * 24 * 14)));
+
+		if (strlen($user_record['timezone'])) {
+			date_default_timezone_set($user_record['timezone']);
+			$a->timezone = $user_record['timezone'];
+		}
+
+		$masterUid = $user_record['uid'];
+
+		if (!empty($_SESSION['submanage'])) {
+			$user = DBA::selectFirst('user', ['uid'], ['uid' => $_SESSION['submanage']]);
+			if (DBA::isResult($user)) {
+				$masterUid = $user['uid'];
+			}
+		}
+
+		$a->identities = User::identities($masterUid);
+
+		if ($login_initial) {
+			$a->getLogger()->info('auth_identities: ' . print_r($a->identities, true));
+		}
+
+		if ($login_refresh) {
+			$a->getLogger()->info('auth_identities refresh: ' . print_r($a->identities, true));
+		}
+
+		$contact = DBA::selectFirst('contact', [], ['uid' => $_SESSION['uid'], 'self' => true]);
+		if (DBA::isResult($contact)) {
+			$a->contact = $contact;
+			$a->cid = $contact['id'];
+			self::set('cid', $a->cid);
+		}
+
+		header('X-Account-Management-Status: active; name="' . $user_record['username'] . '"; id="' . $user_record['nickname'] . '"');
+
+		if ($login_initial || $login_refresh) {
+			DBA::update('user', ['login_date' => DateTimeFormat::utcNow()], ['uid' => $_SESSION['uid']]);
```

Comment:
What about `Contact::updateLogin($uid, $parentUid)`

Reply:
Because it isn't a `Contact` for starters?

Reply:
Ups!!!! I'm sorry!!! I could swear I saw `contact`